qiita_db/processing_job.py

```diff
@@ -6,21 +6,295 @@
+                            self.processes[results['Job_Id']] = results
+                            self.queue.put(results)
+            else:
+                self.queue.put('QUIT')
+                self.event.set()
+                # don't join(), since we are exiting from the main loop
+
+            sleep(self.polling_value)
+
+    def stop(self):
+        # 'poison pill' to thread/process
+        self.queue.put('QUIT')
+        # setting self.event is a safe way of communicating a boolean
+        # value across processes and threads.
+        # when this event is 'set' by the main line of execution in Qiita,
+        # (or in any other process if need be), Watcher's run loop will
+        # stop and the Watcher process will exit.
+        self.event.set()
+        # Here, it is assumed that we are running this from the main
+        # context. By joining(), we're waiting for the Watcher process to
+        # end before returning from this method.
+        self.join()
+
+
+def launch_local(env_script, start_script, url, job_id, job_dir):
+
+    # launch_local() differs from launch_torque(), as no Watcher() is used.
+    # each launch_local() process will execute the cmd as a child process,
+    # wait, and update the database once cmd has completed.
+    #
+    # As processes are lighter weight than jobs, this should be fine.
+    # This is how the current job model works locally.
+
+    cmd = [start_script, url, job_id, job_dir]
+
+    # When Popen() executes, the shell is not in interactive mode,
+    # so it is not sourcing any of the bash configuration files
+    # We need to source it so the env_script are available
+    cmd = "bash -c '%s; %s'" % (env_script, ' '.join(cmd))
+
+    # Popen() may also need universal_newlines=True
+    proc = Popen(cmd, shell=True, stdout=PIPE, stderr=PIPE)
+
+    # Communicate pulls all stdout/stderr from the PIPEs
+    # This call waits until cmd is done
+    stdout, stderr = proc.communicate()
+
+    # proc.returncode will be equal to None if the process hasn't finished
+    # yet. If cmd was terminated by a SIGNAL, it will be a negative value.
+    # (*nix platforms only)
+    error = None
+
+    if proc.returncode != 0:
+        error = "error from launch_local when launching cmd='%s'" % cmd
+        error = "%s\n%s\n%s" % (error, stdout, stderr)
+
+        # Forcing the creation of a new connection
+        qdb.sql_connection.create_new_transaction()
+        ProcessingJob(job_id).complete(False, error=error)
+
+
+def launch_torque(env_script, start_script, url, job_id, job_dir,
+                  dependent_job_id, resource_params):
+
+    # note that job_id is Qiita's UUID, not a Torque job ID
+    cmd = [start_script, url, job_id, job_dir]
+
+    # generating file contents to be used with qsub
+    lines = []
+
+    # TODO: is PBS_JOBID is being set correctly?
+    lines.append("echo $PBS_JOBID")
+
+    # TODO: revisit below
+    lines.append("source ~/.bash_profile")
```

Comment:
This makes the assumption that ~/.bash_profile is necessary and is present in the system, which is not a guarantee, right?

Reply:
It's legacy from the qsub_launcher script. I'd like to leave it as is for now, and revisit it in the future.